File: test/node/disbursement/bank_account_inquiry_test.ts
/**
 * @fileoverview Bank Account Inquiry Test Suite for DANA Disbursement Integration
 * 
 * This suite validates the bank account inquiry functionality of the DANA Disbursement API.
 * Each test case includes comprehensive documentation following JSDoc standards.
 * 
 * @author Integration Test Team
 * @version 1.0.0
 * @since 2024
 * @module BankAccountInquiryTests
 * @requires dana-node
 * @requires uuid
 * @requires dotenv
 */

import Dana from 'dana-node';
import { v4 as uuidv4 } from 'uuid';
import * as path from 'path';
import * as dotenv from 'dotenv';

// Import helper functions and assertion utilities
import { getRequest } from '../helper/util';
import { assertResponse, assertFailResponse } from '../helper/assertion';
import { fail } from 'assert';
import { ResponseError } from 'dana-node';
import { executeManualApiRequest } from '../helper/apiHelpers';

// Load environment variables from .env file
dotenv.config();

// Test configuration constants
const titleCase = "BankAccountInquiry";
const jsonPathFile = path.resolve(__dirname, '../../../resource/request/components/Disbursement.json');

// Initialize DANA SDK client with environment configuration
const dana = new Dana({
  partnerId: process.env.X_PARTNER_ID || '',
  privateKey: process.env.PRIVATE_KEY || '',
  origin: process.env.ORIGIN || '',
  env: process.env.ENV || 'sandbox'
});

/**
 * Bank Account Inquiry Test Suite
 * 
 * @description Comprehensive test coverage for bank account inquiry operations
 * @author Integration Test Team
 * @version 1.0.0
 * @testSuite BankAccountInquiry
 * @component Disbursement
 * @apiVersion v1.0
 */
describe('Disbursement - Bank Account Inquiry Tests', () => {

  /**
   * Test Case: Successful Bank Account Inquiry
   * 
   * @description Validates that a valid bank account inquiry request processes successfully
   * @testId INQUIRY_BANK_ACCOUNT_VALID_DATA_AMOUNT_001
   * @priority High
   * @category Positive Test
   * @expectedResult HTTP 200 with successful bank account inquiry response including account details
   * @prerequisites Valid partner credentials and valid bank account information
   * @author Integration Test Team
   * @since 1.0.0
   */
  test('InquiryBankAccountValidDataAmount - should successfully inquire bank account', async () => {
    const caseName = "InquiryBankAccountValidDataAmount";
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);

    // Assign unique reference for test isolation
    const partnerReferenceNo = uuidv4();
    requestData.partnerReferenceNo = partnerReferenceNo;

    try {
      // Execute bank account inquiry API call
      const response = await dana.disbursementApi.bankAccountInquiry(requestData);
      await assertResponse(jsonPathFile, titleCase, caseName, response, {
              'partnerReferenceNo': partnerReferenceNo
            });
    } catch (e) {
      console.error('Bank account inquiry test failed:', e);
      throw e;
    }
  });

  /**
   * Test Case: Bank Account Inquiry with Insufficient Fund
   * 
   * @description Validates that bank account inquiries with insufficient funds are properly handled
   * @testId INQUIRY_BANK_ACCOUNT_INSUFFICIENT_FUND_002
   * @priority Medium
   * @category Negative Test
   * @expectedResult HTTP 400 with insufficient fund error response
   * @prerequisites Valid partner credentials but insufficient account balance
   * @errorCode INSUFFICIENT_BALANCE
   * @author Integration Test Team
   * @since 1.0.0
   */
  test('InquiryBankAccountInsufficientFund - should fail inquiry due to insufficient fund', async () => {
    const caseName = "InquiryBankAccountInsufficientFund";
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);

    // Assign unique reference for test isolation
    const partnerReferenceNo = uuidv4();
    requestData.partnerReferenceNo = partnerReferenceNo;

    try {
      // Call the widget payment API with the request data
      const response = await dana.disbursementApi.bankAccountInquiry(requestData);
      // Assert the failure response against the expected result
      await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response) , {
        'partnerReferenceNo': partnerReferenceNo
      });
    } catch (e: any) {
      // If a ResponseError occurs, assert the failure response
      if (e instanceof ResponseError) {
        await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse),
                  { 'partnerReferenceNo': partnerReferenceNo });
      } else {
        // If another error occurs, fail the test with the error message
        fail('Payment test failed: ' + (e.message || e));
      }
    }
  });

  /**
   * Test Case: Bank Account Inquiry with Inactive Account
   * 
   * @description Validates that inquiries for inactive bank accounts are properly rejected
   * @testId INQUIRY_BANK_ACCOUNT_INACTIVE_ACCOUNT_003
   * @priority Medium
   * @category Negative Test
   * @expectedResult HTTP 400 with inactive account error response
   * @prerequisites Valid partner credentials but target bank account is inactive
   * @errorCode ACCOUNT_INACTIVE
   * @author Integration Test Team
   * @since 1.0.0
   */
  test('InquiryBankAccountInactiveAccount - should fail inquiry due to inactive account', async () => {
    const caseName = "InquiryBankAccountInactiveAccount";
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);

    // Assign unique reference for test isolation
    const partnerReferenceNo = uuidv4();
    requestData.partnerReferenceNo = partnerReferenceNo;

    try {
      // Call the widget payment API with the request data
      const response = await dana.disbursementApi.bankAccountInquiry(requestData);
      // Assert the failure response against the expected result
      await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response) , {
        'partnerReferenceNo': partnerReferenceNo
      });
    } catch (e: any) {
      // If a ResponseError occurs, assert the failure response
      if (e instanceof ResponseError) {
        await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse),
                  { 'partnerReferenceNo': partnerReferenceNo });
      } else {
        // If another error occurs, fail the test with the error message
        fail('Payment test failed: ' + (e.message || e));
      }
    }
  });

  /**
   * Test Case: Bank Account Inquiry with Invalid Merchant
   * 
   * @description Validates that bank account inquiries with invalid merchant information are rejected
   * @testId INQUIRY_BANK_ACCOUNT_INVALID_MERCHANT_004
   * @priority Medium
   * @category Validation Test
   * @expectedResult HTTP 400 with invalid merchant error response
   * @prerequisites Invalid merchant credentials or information
   * @errorCode INVALID_MERCHANT
   * @author Integration Test Team
   * @since 1.0.0
   */
  test('InquiryBankAccountInvalidMerchant - should fail inquiry due to invalid merchant', async () => {
    const caseName = "InquiryBankAccountInvalidMerchant";
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);

    // Assign unique reference for test isolation
    const partnerReferenceNo = uuidv4();
    requestData.partnerReferenceNo = partnerReferenceNo;

    try {
      // Call the widget payment API with the request data
      const response = await dana.disbursementApi.bankAccountInquiry(requestData);
      // Assert the failure response against the expected result
      await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response) , {
        'partnerReferenceNo': partnerReferenceNo
      });
    } catch (e: any) {
      // If a ResponseError occurs, assert the failure response
      if (e instanceof ResponseError) {
        await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse),
                  { 'partnerReferenceNo': partnerReferenceNo });
      } else {
        // If another error occurs, fail the test with the error message
        fail('Payment test failed: ' + (e.message || e));
      }
    }
  });

  /**
   * Test Case: Bank Account Inquiry with Invalid Card
   * 
   * @description Validates that bank account inquiries with invalid card information are rejected
   * @testId INQUIRY_BANK_ACCOUNT_INVALID_CARD_005
   * @priority Medium
   * @category Validation Test
   * @expectedResult HTTP 400 with invalid card error response
   * @prerequisites Valid partner credentials but invalid card/account number
   * @errorCode INVALID_CARD_NUMBER
   * @author Integration Test Team
   * @since 1.0.0
   */
  test('InquiryBankAccountInvalidCard - should fail inquiry due to invalid card', async () => {
    const caseName = "InquiryBankAccountInvalidCard";
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);

    // Assign unique reference for test isolation
    const partnerReferenceNo = uuidv4();
    requestData.partnerReferenceNo = partnerReferenceNo;

    try {
      // Call the widget payment API with the request data
      const response = await dana.disbursementApi.bankAccountInquiry(requestData);
      // Assert the failure response against the expected result
      await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response) , {
        'partnerReferenceNo': partnerReferenceNo
      });
    } catch (e: any) {
      // If a ResponseError occurs, assert the failure response
      if (e instanceof ResponseError) {
        await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse),
                  { 'partnerReferenceNo': partnerReferenceNo });
      } else {
        // If another error occurs, fail the test with the error message
        fail('Payment test failed: ' + (e.message || e));
      }
    }
  });

  /**
   * Test Case: Bank Account Inquiry with Invalid Field Format
   * 
   * @description Validates that bank account inquiries with invalid field formats are rejected
   * @testId INQUIRY_BANK_ACCOUNT_INVALID_FIELD_FORMAT_006
   * @priority Medium
   * @category Validation Test
   * @expectedResult HTTP 400 with invalid field format error response
   * @prerequisites Valid partner credentials but malformed request fields
   * @errorCode INVALID_FIELD_FORMAT
   * @author Integration Test Team
   * @since 1.0.0
   */
  test('InquiryBankAccountInvalidFieldFormat - should fail inquiry due to invalid field format', async () => {
    const caseName = "InquiryBankAccountInvalidFieldFormat";
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);

    // Assign unique reference for test isolation
    const partnerReferenceNo = uuidv4();
    requestData.partnerReferenceNo = partnerReferenceNo;

    try {
      // Call the widget payment API with the request data
      const response = await dana.disbursementApi.bankAccountInquiry(requestData);
      // Assert the failure response against the expected result
      await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response) , {
        'partnerReferenceNo': partnerReferenceNo
      });
    } catch (e: any) {
      // If a ResponseError occurs, assert the failure response
      if (e instanceof ResponseError) {
        await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse),
                  { 'partnerReferenceNo': partnerReferenceNo });
      } else {
        // If another error occurs, fail the test with the error message
        fail('Payment test failed: ' + (e.message || e));
      }
    }
  });

  /**
   * Test Case: Bank Account Inquiry with Missing Mandatory Field
   * 
   * @description Validates that bank account inquiries missing mandatory fields are rejected
   * @testId INQUIRY_BANK_ACCOUNT_MISSING_MANDATORY_FIELD_007
   * @priority High
   * @category Validation Test
   * @expectedResult HTTP 400 with missing mandatory field error response
   * @prerequisites Valid partner credentials but missing required request fields
   * @errorCode MISSING_MANDATORY_FIELD
   * @author Integration Test Team
   * @since 1.0.0
   */
  test('InquiryBankAccountMissingMandatoryField - should fail inquiry due to missing mandatory field', async () => {
    const caseName = "InquiryBankAccountMissingMandatoryField";
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);

    // Assign unique reference for test isolation
    const partnerReferenceNo = uuidv4();
    requestData.partnerReferenceNo = partnerReferenceNo;

    try {
      // Call the widget payment API with the request data
      const response = await dana.disbursementApi.bankAccountInquiry(requestData);
      // Assert the failure response against the expected result
      await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response) , {
        'partnerReferenceNo': partnerReferenceNo
      });
    } catch (e: any) {
      // If a ResponseError occurs, assert the failure response
      if (e instanceof ResponseError) {
        await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse),
                  { 'partnerReferenceNo': partnerReferenceNo });
      } else {
        // If another error occurs, fail the test with the error message
        fail('Payment test failed: ' + (e.message || e));
      }
    }
  });

  /**
   * Test Case: Bank Account Inquiry with Unauthorized Signature
   * 
   * @description Validates that bank account inquiries with unauthorized signatures are rejected
   * @testId INQUIRY_BANK_ACCOUNT_UNAUTHORIZED_SIGNATURE_008
   * @priority High
   * @category Security Test
   * @expectedResult HTTP 401 with unauthorized signature error response
   * @prerequisites Invalid or tampered signature in request headers
   * @errorCode UNAUTHORIZED_SIGNATURE
   * @author Integration Test Team
   * @since 1.0.0
   */
  test('InquiryBankAccountUnauthorizedSignature - should fail inquiry due to unauthorized signature', async () => {
    const caseName = "InquiryBankAccountUnauthorizedSignature";
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);

    // Assign unique reference for test isolation
    const partnerReferenceNo = uuidv4();
    requestData.partnerReferenceNo = partnerReferenceNo;

    try {
      const baseUrl: string = 'https://api.sandbox.dana.id';
      const apiPath: string = '/v1.0/emoney/bank-account-inquiry.htm';

      const customHeaders: Record<string, string> = {
        'X-SIGNATURE': '85be817c55b2c135157c7e89f52499bf0c25ad6eeebe04a986e8c862561b19a5'
      };

      await executeManualApiRequest(
        caseName,
        "POST",
        baseUrl + apiPath,
        apiPath,
        requestData,
        customHeaders
      );
      fail("Expected an error but the API call succeeded");
    } catch (e: any) {
      // If a ResponseError occurs, assert the failure response
      if (e instanceof ResponseError) {
        await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse), { partnerReferenceNo });
      } else {
        // If another error occurs, fail the test with the error message
        fail('Payment test failed: ' + (e.message || e));
      }
    }
  });
});

File: test/go/disbursement/bank_account_inquiry_test.go
package disbursement_test

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"testing"
	"time"
	"uat-script/helper"

	"github.com/dana-id/dana-go/v2/disbursement/v1"
	"github.com/google/uuid"
)

const (
	bankAccountInquiryTitleCase = "BankAccountInquiry"
	bankAccountInquiryJsonPath  = "../../../resource/request/components/Disbursement.json"
	bankAccountInquiryEndpoint  = "https://api.sandbox.dana.id/v1.0/emoney/bank-account-inquiry.htm"
	bankAccountInquiryPath      = "/v1.0/emoney/bank-account-inquiry.htm"
)

// assertBankAccountInquiryErrorViaSDK calls the SDK with the fixture payload and logs the error response.
func assertBankAccountInquiryErrorViaSDK(t *testing.T, caseName string) error {
	t.Helper()

	jsonDict, err := helper.GetRequest(bankAccountInquiryJsonPath, bankAccountInquiryTitleCase, caseName)
	if err != nil {
		return fmt.Errorf("Failed to get request data: %v", err)
	}

	partnerReferenceNo := uuid.New().String()
	jsonDict["partnerReferenceNo"] = partnerReferenceNo

	bankAccountInquiryRequest := &disbursement.BankAccountInquiryRequest{}
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		return fmt.Errorf("Failed to marshal JSON: %v", err)
	}
	if err = json.Unmarshal(jsonBytes, bankAccountInquiryRequest); err != nil {
		return fmt.Errorf("Failed to unmarshal JSON: %v", err)
	}

	ctx := context.Background()
	apiResponse, httpResponse, err := helper.ApiClient.DisbursementAPI.BankAccountInquiry(ctx).BankAccountInquiryRequest(*bankAccountInquiryRequest).Execute()
	if httpResponse != nil {
		defer httpResponse.Body.Close()
	}

	fmt.Printf("case=%s err=%v apiResponse=%+v\n", caseName, err, apiResponse)
	if httpResponse != nil && httpResponse.Body != nil {
		if body, readErr := io.ReadAll(httpResponse.Body); readErr == nil {
			fmt.Printf("case=%s httpBody=%s\n", caseName, string(body))
		}
	}
	return nil
}

func TestInquiryBankAccountValidDataAmount(t *testing.T) {
	helper.RetryTest(t, 3, 2*time.Second, func() error {
		caseName := "InquiryBankAccountValidDataAmount"

		// Get the request data from the JSON file
		jsonDict, err := helper.GetRequest(bankAccountInquiryJsonPath, bankAccountInquiryTitleCase, caseName)
		if err != nil {
			t.Fatalf("Failed to get request data: %v", err)
		}

		// Set the correct partner reference number
		var partnerReferenceNo = uuid.New().String()
		jsonDict["partnerReferenceNo"] = partnerReferenceNo

		// Create the BankAccountInquiryRequest object and populate it with JSON data
		bankAccountInquiryRequest := &disbursement.BankAccountInquiryRequest{}
		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			t.Fatalf("Failed to marshal JSON: %v", err)
		}

		err = json.Unmarshal(jsonBytes, bankAccountInquiryRequest)
		if err != nil {
			t.Fatalf("Failed to unmarshal JSON: %v", err)
		}

		// Make the API call
		ctx := context.Background()
		apiResponse, httpResponse, err := helper.ApiClient.DisbursementAPI.BankAccountInquiry(ctx).BankAccountInquiryRequest(*bankAccountInquiryRequest).Execute()
		if err != nil {
			fmt.Printf("[REF] case=%s partnerReferenceNo=%s\n", caseName, partnerReferenceNo)
			t.Fatalf("API call failed: %v", err)
		}
		defer httpResponse.Body.Close()

		// Convert the response to JSON for assertion
		responseJSON, err := apiResponse.MarshalJSON()
		if err != nil {
			t.Fatalf("Failed to convert response to JSON: %v", err)
		}

		// Create variable dictionary for dynamic values
		variableDict := map[string]interface{}{
			"partnerReferenceNo": partnerReferenceNo,
		}

		// Assert the API response with variable substitution
		err = helper.AssertResponse(bankAccountInquiryJsonPath, bankAccountInquiryTitleCase, caseName, string(responseJSON), variableDict)
		if err != nil {
			t.Fatal(err)
		}
		return nil
	})
}

func TestInquiryBankAccountInsufficientFund(t *testing.T) {
	helper.RetryTest(t, 3, 2*time.Second, func() error {
		return assertBankAccountInquiryErrorViaSDK(t, "InquiryBankAccountInsufficientFund")
	})
}

func TestInquiryBankAccountInactiveAccount(t *testing.T) {
	helper.RetryTest(t, 3, 2*time.Second, func() error {
		return assertBankAccountInquiryErrorViaSDK(t, "InquiryBankAccountInactiveAccount")
	})
}

func TestInquiryBankAccountInvalidMerchant(t *testing.T) {
	helper.RetryTest(t, 3, 2*time.Second, func() error {
		return assertBankAccountInquiryErrorViaSDK(t, "InquiryBankAccountInvalidMerchant")
	})
}

func TestInquiryBankAccountInvalidCard(t *testing.T) {
	helper.RetryTest(t, 3, 2*time.Second, func() error {
		return assertBankAccountInquiryErrorViaSDK(t, "InquiryBankAccountInvalidCard")
	})
}

func TestInquiryBankAccountInvalidFieldFormat(t *testing.T) {
	helper.RetryTest(t, 3, 2*time.Second, func() error {
		caseName := "InquiryBankAccountInvalidFieldFormat"

		// Get the request data from the JSON file
		jsonDict, err := helper.GetRequest(bankAccountInquiryJsonPath, bankAccountInquiryTitleCase, caseName)
		if err != nil {
			t.Fatalf("Failed to get request data: %v", err)
		}

		// Set the correct partner reference number
		var partnerReferenceNo = uuid.New().String()
		jsonDict["partnerReferenceNo"] = partnerReferenceNo

		// Create the BankAccountInquiryRequest object and populate it with JSON data
		bankAccountInquiryRequest := &disbursement.BankAccountInquiryRequest{}
		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			t.Fatalf("Failed to marshal JSON: %v", err)
		}

		err = json.Unmarshal(jsonBytes, bankAccountInquiryRequest)
		if err != nil {
			t.Fatalf("Failed to unmarshal JSON: %v", err)
		}

		// Make the API call
		ctx := context.Background()
		_, httpResponse, err := helper.ApiClient.DisbursementAPI.BankAccountInquiry(ctx).BankAccountInquiryRequest(*bankAccountInquiryRequest).Execute()
		if err != nil {
			// Assert the API error response
			err = helper.AssertFailResponse(bankAccountInquiryJsonPath, bankAccountInquiryTitleCase, caseName, httpResponse, map[string]interface{}{
				"partnerReferenceNo": jsonDict["partnerReferenceNo"],
			})
			if err != nil {
				t.Fatal(err)
			}
		} else {
			httpResponse.Body.Close()
			t.Fatal("Expected error but got successful response")
		}
		return nil
	})
}

func TestInquiryBankAccountMissingMandatoryField(t *testing.T) {
	helper.RetryTest(t, 3, 2*time.Second, func() error {
		return assertBankAccountInquiryErrorViaSDK(t, "InquiryBankAccountMissingMandatoryField")
	})
}

func TestInquiryBankAccountUnauthorizedSignature(t *testing.T) {
	helper.RetryTest(t, 3, 2*time.Second, func() error {
		caseName := "InquiryBankAccountUnauthorizedSignature"

		// Get the request data from the JSON file
		jsonDict, err := helper.GetRequest(bankAccountInquiryJsonPath, bankAccountInquiryTitleCase, caseName)
		if err != nil {
			t.Fatalf("Failed to get request data: %v", err)
		}

		// Set the correct partner reference number
		var partnerReferenceNo = uuid.New().String()
		jsonDict["partnerReferenceNo"] = partnerReferenceNo

		// Create the BankAccountInquiryRequest object and populate it with JSON data
		bankAccountInquiryRequest := &disbursement.BankAccountInquiryRequest{}
		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			t.Fatalf("Failed to marshal JSON: %v", err)
		}

		err = json.Unmarshal(jsonBytes, bankAccountInquiryRequest)
		if err != nil {
			t.Fatalf("Failed to unmarshal JSON: %v", err)
		}

		ctx := context.Background()
		customHeaders := map[string]string{
			"X-SIGNATURE": "invalid_signature",
		}
		variableDict := map[string]interface{}{
			"partnerReferenceNo": partnerReferenceNo,
		}

		_ = helper.ExecuteAndAssertErrorResponse(
			t,
			ctx,
			bankAccountInquiryRequest,
			"POST",
			bankAccountInquiryEndpoint,
			bankAccountInquiryPath,
			bankAccountInquiryJsonPath,
			bankAccountInquiryTitleCase,
			caseName,
			customHeaders,
			variableDict,
		)
		return nil
	})
}
